File: 994-rotting-oranges/994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        Queue<int[]>q=new LinkedList<>();
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2){
                    q.offer(new int[]{i,j}); // q= (0,0)
                }
            }
        }
        int[] dx={-1,0,0,1};
        int[] dy={0,1,-1,0};
        int fresh=-1;
        while(!q.isEmpty()){
            int s=q.size();
            for(int j=0;j<s;j++){
            int[] temp=q.poll();
            int x=temp[0];//x=0
            int y=temp[1];//y=0
            
            for(int i=0;i<4;i++){
                int dirx=x+dx[i];//0,1
                int diry=y+dy[i];
                if(dirx<0||diry<0||dirx>=n||diry>=m||grid[dirx][diry]!=1){
                    continue;
                }
                
                    q.offer(new int[]{dirx,diry});
                    grid[dirx][diry]=2;
                
                
            }
                
            }
            fresh++;
            
            
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    return -1;
                }
            }
        }
        if(fresh==-1){
            return 0;
        }
        return fresh;
        
        
    }
}

File: 994-rotting-oranges/994-rotting-oranges.cpp
class Solution {
public:
    //2 1 1
    //1 1 0
    //0 1 1
    //0 0 0
    //2 1 1
    // (0,0),(4,0),(0,1),(1,0)
    int orangesRotting(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        queue<pair<int,int>>q;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2){
                    q.push({i,j});
                }
            }
        }
        int dx[4]={0,0,1,-1};
        int dy[4]={1,-1,0,0};
        int c=-1;
        while(!q.empty()){// =(0,0)
            
           
            int size=q.size();//1
            
            while(size--){//1
             int x=q.front().first;//x=0
            int y=q.front().second;//y=0
                q.pop();
            for(int i=0;i<4;i++){
                int ix=x+dx[i];//ix=0
                int iy=y+dy[i];//iy=2
                if(ix<0||iy<0||ix>=n||iy>=m||grid[ix][iy]!=1){
                    continue;
                }
               
                q.push({ix,iy});// (0,2),(1,2)
                 grid[ix][iy]=2;
               
            }
                
            }
            c++;//c=4
            
        }
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    return -1;
                }
            }
        }
        if(c==-1){
            return 0;
        }
        return c;
        
        
    }
};